### ISSUE TITLE
src/http_ext.c: uninitialized variable suspected by gcc-14

### ISSUE THREAD

**Issue by chipitsine** [haproxy/haproxy#2481]
### Tool Name and Version

gcc 14

### Code Report

```plain
on Fedora Rawhide


        make CC=gcc ERR=1 \
          TARGET=linux-glibc \
          USE_OPENSSL=1 USE_LUA=1 \
          DEBUG_CFLAGS="-m32" \
          LDFLAGS="-m32" 

```

```
  CC      src/mqtt.o
src/http_ext.c: In function 'sample_conv_7239_n2nn':
src/http_ext.c:1830:50: error: '*(struct ist *)((char *)&ctx + offsetof(struct forwarded_header_node, nodename.<U91c8>)).len' may be used uninitialized [-Werror=maybe-uninitialized]
 1830 |                                 smp->data.u.ipv4 = ((struct sockaddr_in *)&ctx.nodename.ip)->sin_addr;
      |                                 ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated due to -Wfatal-errors.
cc1: all warnings being treated as errors
make: *** [Makefile:1006: src/http_ext.o] Error 1
make: *** Waiting for unfinished jobs....
Error: Process completed with exit code 2.
```
```


### Additional Information

_No response_

### Output of `haproxy -vv`

```plain
HAProxy version 3.0-dev4-f36208a959c8 2024/03/07 - https://haproxy.org/
Status: development branch - not safe for use in production.
Known bugs: https://github.com/haproxy/haproxy/issues?q=is:issue+is:open
Running on: Linux 6.5.0-1015-azure #15~22.04.1-Ubuntu SMP Tue Feb 13 01:15:12 UTC 2024 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -m32 -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_OPENSSL=1 USE_LUA=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_AWSLC -OPENSSL_WOLFSSL -OT -PCRE -PCRE2 -PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL -PROMEX -PTHREAD_EMULATION -QUIC -QUIC_OPENSSL_COMPAT +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 -SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=128, default=4).
Built with OpenSSL version : OpenSSL 3.2.1 30 Jan 2024
Running on OpenSSL version : OpenSSL 3.2.1 30 Jan 2024
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.4.6
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built without PCRE or PCRE2 support (using libc's regex instead)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 14.0.1 20240228 (Red Hat 14.0.1-0)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : none

Available filters :
	[BWLIM] bwlim-in
	[BWLIM] bwlim-out
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[SPOE] spoe
	[TRACE] trace
```

**Comment by Darlelet** [comment]
Thanks @chipitsine 

Here is the same report but with `forwarded_header_nodename` using a struct instead of an union for its members (the union makes the report a bit confusing):
```diff
diff --git a/include/haproxy/http_ext-t.h b/include/haproxy/http_ext-t.h
index 68eb047e2..3202d18d5 100644
--- a/include/haproxy/http_ext-t.h
+++ b/include/haproxy/http_ext-t.h
@@ -34,7 +34,7 @@ enum forwarded_header_attribute_type {
 };
 
 struct forwarded_header_nodename {
-	union {
+	struct {
 		struct sockaddr_storage ip;
 		struct ist obfs;
 	};
```

```
src/http_ext.c: In function ‘sample_conv_7239_n2nn’:
src/http_ext.c:1830:50: warning: ‘*(unsigned int *)((char *)&ctx + offsetof(struct forwarded_header_node, nodename.<Ufc78>.ip.__ss_padding[2]))’ may be used uninitialized [-Wmaybe-uninitialized]
 1830 |                                 smp->data.u.ipv4 = ((struct sockaddr_in *)&ctx.nodename.ip)->sin_addr;
      |                                 ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/http_ext.c:1806:38: note: ‘*(unsigned int *)((char *)&ctx + offsetof(struct forwarded_header_node, nodename.<Ufc78>.ip.__ss_padding[2]))’ was declared here
 1806 |         struct forwarded_header_node ctx;
      |                                      ^~~
At top level:
cc1: note: unrecognized command-line option ‘-Wno-atomic-alignment’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-string-plus-int’ may have been intended to silence earlier diagnostics
```

Looks like it's due to padding involved in `sockaddr_storage` struct when doing a 32bits build since the `ctx` struct isn't zeroed out (the warning isn't produced when building without `-m32`). Considering gcc-14 is still under development, I can't tell if this should be considered as a false positive (hoping that it would be silenced before gcc-14 is released), or if we should do something on our side to suppress the warning. It doesn't look like a bug to me, but I can be wrong.

This should help to silence the warning (if we consider that there is no issue):
```diff
diff --git a/src/http_ext.c b/src/http_ext.c
index 269a97f5e..a0fa46854 100644
--- a/src/http_ext.c
+++ b/src/http_ext.c
@@ -1803,7 +1803,7 @@ static int sample_conv_7239_field(const struct arg *args, struct sample *smp, vo
 static int sample_conv_7239_n2nn(const struct arg *args, struct sample *smp, void *private)
 {
 	struct ist input = ist2(smp->data.u.str.area, smp->data.u.str.data);
-	struct forwarded_header_node ctx;
+	struct forwarded_header_node ctx = { }; // shut GCC warning
 	struct buffer *output;
 
 	if (http_7239_extract_node(&input, &ctx, 1) == 0)
```

**Comment by wtarreau** [comment]
I suspect this is caused by the (void*) casts in http_7239_extract_nodename() that are causing a breakage of aliasing and preventing the compiler from knowing if the fields are entirely set or only partially. Someone with gcc-14 might want to test this:
```diff
diff --git a/src/http_ext.c b/src/http_ext.c
index 269a97f5e..4c5553672 100644
--- a/src/http_ext.c
+++ b/src/http_ext.c
@@ -256,7 +256,7 @@ static inline int http_7239_extract_nodename(struct ist *input, struct forwarded
 
                /* ipv6 address */
                if (nodename) {
-                       struct sockaddr_in6 *addr = (void *)&nodename->ip;
+                       struct sockaddr_in6 *addr = (struct sockaddr_in6 *)&nodename->ip;
 
                        ip6 = &addr->sin6_addr;
                        addr->sin6_family = AF_INET6;
@@ -269,7 +269,7 @@ static inline int http_7239_extract_nodename(struct ist *input, struct forwarded
 
                /* ipv4 address */
                if (nodename) {
-                       struct sockaddr_in *addr = (void *)&nodename->ip;
+                       struct sockaddr_in *addr = (struct sockaddr_in *)&nodename->ip;
 
                        ip = &addr->sin_addr;
                        addr->sin_family = AF_INET;
```

**Comment by chipitsine** [comment]
```
   CC      src/mqtt.o
src/http_ext.c: In function 'sample_conv_7239_n2nn':
src/http_ext.c:1830:50: error: '*(struct ist *)((char *)&ctx + offsetof(struct forwarded_header_node, nodename.<U41c8>)).len' may be used uninitialized [-Werror=maybe-uninitialized]
 1830 |                                 smp->data.u.ipv4 = ((struct sockaddr_in *)&ctx.nodename.ip)->sin_addr;
      |                                 ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated due to -Wfatal-errors.
cc1: all warnings being treated as errors
make: *** [Makefile:1006: src/http_ext.o] Error 1
make: *** Waiting for unfinished jobs....
Error: Process completed with exit code 2.
```

**Comment by wtarreau** [comment]
Thanks for the test Ilya. No idea for now :-(

**Comment by Darlelet** [comment]
Ok so I found some time to run a few tests with gcc-14 on fedora 40: I'm pretty sure it is a false positive now.

It simply looks like gcc is loosing the track in `http_7239_extract_ipv4()`. If I un-inline the function or if I bypass the function by doing the same operations directly in the calling function (`http_7239_extract_nodename()`), gcc seems happy.. :)

While at it, I noticed that when `<ip>` is provided by the caller in `http_7239_extract_{ipv4,ipv6}()` we perform an useless copy (we can directly use `<ip>` as `inet_pton()` dst argument in this case).

Below is a patch to skip this extra copy. By chance, this also helps to silence the warning as gcc doesn't need to track 
`buf` usage anymore.

[0001-OPTIM-http_ext-avoid-useless-copy-in-http_7239_extra.patch.txt](https://github.com/haproxy/haproxy/files/14638230/0001-OPTIM-http_ext-avoid-useless-copy-in-http_7239_extra.patch.txt)
